In [27]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
from sklearn.ensemble import  RandomForestClassifier
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, precision_score, recall_score
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression
import xgboost as xgb
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingClassifier

In [3]:
# os.chdir('drive/My Drive/Datasets')
os.listdir('.')

['heart.csv',
 'iris.csv',
 'pima-indians-diabetes.csv',
 'sonar.csv',
 'housing.csv',
 'winequality-white.csv',
 'auto.csv']

In [4]:
dataset = pd.read_csv('sonar.csv')

In [5]:
dataset.info()
dataset.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      208 non-null    float64
 1   V2      208 non-null    float64
 2   V3      208 non-null    float64
 3   V4      208 non-null    float64
 4   V5      208 non-null    float64
 5   V6      208 non-null    float64
 6   V7      208 non-null    float64
 7   V8      208 non-null    float64
 8   V9      208 non-null    float64
 9   V10     208 non-null    float64
 10  V11     208 non-null    float64
 11  V12     208 non-null    float64
 12  V13     208 non-null    float64
 13  V14     208 non-null    float64
 14  V15     208 non-null    float64
 15  V16     208 non-null    float64
 16  V17     208 non-null    float64
 17  V18     208 non-null    float64
 18  V19     208 non-null    float64
 19  V20     208 non-null    float64
 20  V21     208 non-null    float64
 21  V22     208 non-null    float64
 22  V2

V1      V2      V3      V4  ...     V58     V59     V60  Class
0  0.0200  0.0371  0.0428  0.0207  ...  0.0084  0.0090  0.0032      1
1  0.0453  0.0523  0.0843  0.0689  ...  0.0049  0.0052  0.0044      1
2  0.0262  0.0582  0.1099  0.1083  ...  0.0164  0.0095  0.0078      1
3  0.0100  0.0171  0.0623  0.0205  ...  0.0044  0.0040  0.0117      1
4  0.0762  0.0666  0.0481  0.0394  ...  0.0048  0.0107  0.0094      1

[5 rows x 61 columns]

In [6]:
dataset.corr()

V1        V2        V3  ...       V59       V60     Class
V1     1.000000  0.735896  0.571537  ...  0.357116  0.347078 -0.271694
V2     0.735896  1.000000  0.779916  ...  0.352200  0.358761 -0.231238
V3     0.571537  0.779916  1.000000  ...  0.425047  0.373948 -0.192195
V4     0.491438  0.606684  0.781786  ...  0.420266  0.400626 -0.250638
V5     0.344797  0.419669  0.546141  ...  0.290982  0.253710 -0.222232
...         ...       ...       ...  ...       ...       ...       ...
V57    0.313725  0.280341  0.380819  ...  0.431295  0.287219 -0.000933
V58    0.368132  0.353042  0.334108  ...  0.550235  0.329827 -0.184191
V59    0.357116  0.352200  0.425047  ...  1.000000  0.642872 -0.130826
V60    0.347078  0.358761  0.373948  ...  0.642872  1.000000 -0.090055
Class -0.271694 -0.231238 -0.192195  ... -0.130826 -0.090055  1.000000

[61 rows x 61 columns]

In [26]:
# sb.set_style("whitegrid");
# sb.pairplot(dataset, hue="target", height=3);
# plt.show()

In [9]:
feature = dataset.drop(['Class'], axis=1) #outcome tidak dibutuhkan
label = dataset['Class'] #target prediksi

In [10]:
scaler = preprocessing.StandardScaler()

feature_scaled = scaler.fit_transform(feature.values) 

In [11]:
feature_train, feature_test, label_train, label_test = train_test_split(feature_scaled, label, test_size=0.2, random_state=10)

In [12]:
#Random Forest Classification
random_forest = RandomForestClassifier(n_estimators=800, bootstrap=True, max_features = "sqrt")
random_forest.fit(feature_train, label_train)

predictions = random_forest.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Testing Set: {:.3f}'.format(random_forest.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = random_forest.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(random_forest.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Testing Set: 0.738
f1 score
1.0
precision score
1.0
recall score
1.0

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.738
f1 score
0.7379466817923994
precision score
0.7611764705882353
recall score
0.7569444444444444


In [15]:
#k-NN
knn = KNeighborsClassifier(19)
knn.fit(feature_train, label_train)

predictions = knn.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Testing Set: {:.3f}'.format(knn.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = knn.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(knn.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Testing Set: 0.571
f1 score
0.7576833550321537
precision score
0.8123188405797102
recall score
0.7539401973781117

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.571
f1 score
0.5625
precision score
0.625
recall score
0.6041666666666667


In [28]:
#GradienBoosting
GB_classi = GradientBoostingClassifier(n_estimators=50)
GB_classi.fit(feature_train, label_train)


predictions = GB_classi.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(GB_classi.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = GB_classi.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(GB_classi.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 1.000
f1 score
1.0
precision score
1.0
recall score
1.0

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.738
f1 score
0.7379466817923994
precision score
0.7611764705882353
recall score
0.7569444444444444


In [17]:
#SVM
svm_classifier = SVC(kernel='linear')
svm_classifier.fit(feature_train, label_train)

predictions = svm_classifier.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(svm_classifier.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = svm_classifier.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(svm_classifier.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 0.952
f1 score
0.9513553113553113
precision score
0.9498030059827812
recall score
0.9540433053468846

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.786
f1 score
0.7855927396483267
precision score
0.8105882352941176
recall score
0.8055555555555556


In [18]:
#Linear SVC
classifier_linear_svc = SVC()
classifier_linear_svc.fit(feature_train, label_train)

predictions = classifier_linear_svc.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(classifier_linear_svc.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = classifier_linear_svc.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(classifier_linear_svc.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 0.988
f1 score
0.9877743408454853
precision score
0.9877743408454853
recall score
0.9877743408454853

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.833
f1 score
0.8332387975042541
precision score
0.86
recall score
0.8541666666666667


In [19]:
#XGBoost
xgb_model = xgb.XGBClassifier(objective="reg:squarederror")
xgb_model.fit(feature_train, label_train)

predictions = xgb_model.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(xgb_model.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = xgb_model.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(xgb_model.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 1.000
f1 score
1.0
precision score
1.0
recall score
1.0

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.667
f1 score
0.6659090909090909
precision score
0.6947115384615384
recall score
0.6875


In [20]:
#Gaussian Process Classifier
gaus_pc = GaussianProcessClassifier() 
gaus_pc.fit(feature_train, label_train)

predictions = gaus_pc.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(gaus_pc.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = gaus_pc.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(gaus_pc.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 1.000
f1 score
1.0
precision score
1.0
recall score
1.0

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.810
f1 score
0.8095238095238096
precision score
0.8263888888888888
recall score
0.8263888888888888


In [22]:
#MLP Classifier
mlp_classi = MLPClassifier(hidden_layer_sizes=(3,), activation='logistic',
                       solver='adam', alpha=0.0001,learning_rate='constant', 
                      learning_rate_init=0.3)
mlp_classi.fit(feature_train, label_train)

predictions = mlp_classi.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(mlp_classi.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = mlp_classi.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(mlp_classi.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 0.922
f1 score
0.9208567132431145
precision score
0.9195906432748537
recall score
0.9227426719693621

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.786
f1 score
0.7826336975273145
precision score
0.7814645308924485
recall score
0.7847222222222222


In [23]:
#ADA Boost
ada_boost = AdaBoostClassifier(n_estimators=250, random_state=15)
ada_boost.fit(feature_train, label_train)

predictions = ada_boost.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(ada_boost.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = ada_boost.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(ada_boost.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 1.000
f1 score
1.0
precision score
1.0
recall score
1.0

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.690
f1 score
0.690300623936472
precision score
0.6990846681922197
recall score
0.7013888888888888


In [24]:
#DecisionTreeClassifier
dec_tc = DecisionTreeClassifier()

dec_tc.fit(feature_train, label_train)

predictions = dec_tc.predict(feature_train)

print("prediksi menggunakan training set")
print('Akurasi dari Training Set: {:.3f}'.format(dec_tc.score(feature_train, label_train))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_train, predictions, average='macro'))
print("precision score")
print(precision_score(label_train, predictions, average='macro'))
print("recall score")
print(recall_score(label_train, predictions, average='macro'))

predictions = dec_tc.predict(feature_test)

print("\nprediksi menggunakan testing set")
print('Akurasi dari Testing Set: {:.3f}'.format(dec_tc.score(feature_test, label_test))) # .score() -> Accuracy
print("f1 score")
print(f1_score(label_test, predictions, average='macro'))
print("precision score")
print(precision_score(label_test, predictions, average='macro'))
print("recall score")
print(recall_score(label_test, predictions, average='macro'))

prediksi menggunakan training set
Akurasi dari Training Set: 1.000
f1 score
1.0
precision score
1.0
recall score
1.0

prediksi menggunakan testing set
Akurasi dari Testing Set: 0.667
f1 score
0.6666666666666666
precision score
0.6805555555555556
recall score
0.6805555555555556


Untuk data sonar.csv Nilai Akurasi yang paling tinggi didapatkan dengan menggunakan algoritma LinearSVC dengan nilai Akurasi : 0.833